pallets/parachain-staking/src/lib.rs

```diff
@@ -2061,85 +2068,162 @@ pub mod pallet {
 					}
 				});
-				TopCandidates::<T>::put(top_candidates);
+				Some((Some(i), top_candidates))
+			} else if top_candidates.try_insert_replace(new_stake.clone()).is_ok() {
+				// case 2: candidate ascends into TopCandidates with new stake
+				// and might replace another candidate if TopCandidates is full
+				Self::deposit_event(Event::EnteredTopCandidates(candidate));
+				Some((None, top_candidates))
+			} else {
+				// case 3: candidate neither was nor will be member of TopCandidates
+				None
+			};
 
-				// update total amount at stake only if candidate is collator
-				if i.saturated_into::<u32>() < MaxSelectedCandidates::<T>::get() {
-					TotalCollatorStake::<T>::mutate(|total| {
-						total.collators = total.collators.saturating_sub(old_self).saturating_add(new_self);
-						total.delegators = total
-							.delegators
-							.saturating_sub(old_delegators)
-							.saturating_add(new_delegators);
-					});
-				}
-			// case 2: candidate newly joins TopCandidates and either replaces
-			// another candidate in the list or updates their own stake
-			} else if let Ok(drop_out) = top_candidates.try_insert_replace(Stake {
-				owner: candidate.clone(),
-				amount: new_total,
-			}) {
-				// retrieve position of candidate in TopCandidates
-				if let Ok(i) = top_candidates.linear_search(&Stake {
-					owner: candidate.clone(),
-					amount: new_total,
-				}) {
-					// update TotalStake only if candidate becomes collator
-					let max_selected_candidates = MaxSelectedCandidates::<T>::get().saturated_into::<usize>();
-					if i < max_selected_candidates {
-						// candidate either pushed out the least staked collator or updated their own
-						// stake
-						let dropped_col_index = max_selected_candidates.min(top_candidates.len().saturating_sub(1));
+			// update storage for TotalCollatorStake and TopCandidates
+			if let Some((maybe_old_idx, top_candidates)) = working_title {
+				let max_selected_candidates = MaxSelectedCandidates::<T>::get().saturated_into::<usize>();
+				let was_collating = maybe_old_idx.map(|i| i < max_selected_candidates).unwrap_or(false);
+				let is_collating = top_candidates
+					.linear_search(&new_stake)
+					.map(|i| i < max_selected_candidates)
+					.unwrap_or(false);
+
+				// only update TopCollatorStake storageif candidate was or will be a collator
+				match (was_collating, is_collating) {
+					(true, true) => {
+						Self::update_total_stake_by(new_self, new_delegators, old_self, old_delegators);
+					}
+					(true, false) => {
+						// candidate was displaced out of collators the current last collator with index
+						// max_selected_candidates - 1
+						let i = max_selected_candidates.min(top_candidates.len()).saturating_sub(1);
+
+						// get displacer
+						let (add_collators, add_delegators) = Self::get_top_candidate_stake_at(&top_candidates, i)
+							// shouldn't be possible to fail, but we handle it gracefully
+							.unwrap_or((new_self, new_delegators));
```

Comment:
I think we can fail if `top_candidates < max_selected_candidates`, right? Kind of an out-of-bound error. So I think that returning `(new_self, new_delegators)` is the right default here.

Reply:
Even that should be covered because our replacement index is min(top_candidates.len(), max_selected_candidates - 1). It can only fail if the candidate at the replacement index does not have an entry in `CandidatePool` which should be impossible.